lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -435,6 +435,15 @@ Value LayoutPropagation::getValueAs(Value value, Attribute encoding) {
       return rewrittenValue;
     OpBuilder rewriter(value.getContext());
     rewriter.setInsertionPointAfterValue(rewrittenValue);
+    // Workaround: The pipeliner will insert async.wait after a pipelined loop
+    // to ensure that there is no pending copies and it is safe to re-use shared
+    // memory. We shouldn't insert ops that may use shared memory in between the
+    // loop and the async copy. This is a hack until we fix the IR
```

Comment:
> We shouldn't insert ops that may use shared memory in between the
>     // loop and the async copy

in between the loop and the async wait?

Reply:
I don't understand why though.

```
scf.for{
}
ttg.convert_layout
async_wait
```
why convert_layout is not allowed?

Reply:
because the `async.wait` is there to make sure there is no more pending copies to shared memory and we then assume that we can re-use the shared memory from the dot. Inserting an instruction in between the for and the async will break this assumption and if the allocation decides to re-use the shared memory for the convert then we have a race condition with the dummy copies from the last loop iterations.

Reply:
It could be fixed by expand the life range of the dot operand to the corresponding async_wait

Reply:
Interesting, you mean extending the liverange of the alloc? We would need to know what alloc the async corresponds to right? 
That would require changing the IR to keep track of this I assume. 
I need to look at this, do you think it is a simple fix?

Reply:
Yes, it has to be marked in the pipeline pass, may need to change the semantics of async_wait. No worry about it for now I think

Reply:
that was my thought too, we probably need to change the async_wait in some way